File: Java/src/session2/td/agregation/NatCalculantAvecDesInts.java
package session2.td.agregation;

import session1.td.FabriqueNaturels;

public class NatCalculantAvecDesInts extends NatDeleguantEtat {

    public static final FabriqueNaturels<Nat> FAB = new NatCalculantAvecDesInts(new IntPositif(1)) ;

    public NatCalculantAvecDesInts(EtatNaturelPur etat) {
        super(etat);
    }

    @Override
    public Nat creerNatAvecEtat(EtatNaturelPur e) {
        return new NatCalculantAvecDesInts(e);
    }

    @Override
    public Nat somme(Nat x) {
        return this.creerNatAvecValeur(this.val() + x.val());
    }

    @Override
    public Nat un() {
        return this.creerNatAvecValeur(1);
    }

    @Override
    public Nat produit(Nat x) {
        return this.creerNatAvecValeur(this.val() * x.val());
    }


    @Override
    public Nat modulo(Nat x) {
        return this.creerNatAvecValeur(this.val()%x.val());
    }

    @Override
    public Nat div(Nat x) {
        return this.creerNatAvecValeur(this.val()/x.val());
    }

    @Override
    public Nat zero() {
        return this.creerZero();
    }

    @Override
    public String toString() {
        return ""+this.etat().toString();
    }
    @Override
    public boolean equals(Object obj) {
        if(!(obj instanceof Nat))
            return false;
        Nat x = (Nat)obj;
        return this.val() == x.val();
    }
}

File: TypeScript/src/session2.td/agregation/NatCalculantAvecDesInts.ts
import {NatDeleguantEtat} from "./NatDeleguantEtat";
import {EtatNaturelPur, EtatNaturelPur_default} from "./EtatNaturelPur";
import {Nat} from "./Nat";
import {FabriqueNaturels} from "../../session1.td/FabriqueNaturels";
import {IntPositif} from "./IntPositif";

export class NatCalculantAvecDesInts extends NatDeleguantEtat {

    public static  FAB : FabriqueNaturels<Nat>  = new NatCalculantAvecDesInts(new IntPositif(1)) ;

    public constructor(etat : EtatNaturelPur) {
        super(etat);
    }

    public creerNatAvecEtat(e : EtatNaturelPur) : Nat{
        return new NatCalculantAvecDesInts(e);
    }

    public somme(x : Nat) : Nat{
        return this.creerNatAvecValeur(this.val() + x.val());
    }

    public un() : Nat{
        return this.creerNatAvecValeur(1);
    }

    public produit(x : Nat) : Nat{
        return this.creerNatAvecValeur(this.val() * x.val());
    }

    public modulo(x : Nat) : Nat{
        return this.creerNatAvecValeur(this.val()%x.val());
    }

    public div(x : Nat) : Nat{
        return this.creerNatAvecValeur(this.val()/x.val());
    }

    public zero() : Nat{
        return this.creerZero();
    }

    public toString() : string{
        return ""+this.etat();
    }

    public equals(obj : Nat) : boolean{
        return this.val() == obj.val();
    }
}